Support capacity scheduling with maskout value specification

Haichao-Zhang opened the pull request:
This PR adds supports to maskout value specification in capacity scheduling.

Haichao-Zhang committed:
Support capacity scheduling with maskout value specification

Haichao-Zhang committed:
Update

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -380,6 +428,10 @@ def add_param_group(self, param_group):
         else:
             super(NewCls, self).add_param_group(param_group)
 
+        capacity_ratio = self._capacity_ratio
```
capacity_ratio should be from scheduler.

Haichao-Zhang replied:
Changed the name to initial_capacity_ratio. The reason is that we need to record initial capacity ratio in order to be used in ``add_param_group``, since  the capacity scheduler (``self._capacity_ratio``) is inaccessible in ``add_param_group``.

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -380,6 +428,10 @@ def add_param_group(self, param_group):
 
+        capacity_ratio = self._capacity_ratio
+        states = self._generate_capacity_mask(capacity_ratio)
+        self._apply_capacity_mask(capacity_ratio, states)
```
why are these additional self._apply_capacity_mask needed for parvi_step?

Haichao-Zhang replied:
Not sure if I understand your question correctly. This is in ``add_param_group``, not in ``parvi_step``. 
The reason to ``_apply_capacity_mask`` in ``add_param_group`` is to correctly handle the parameter values before any forward step.

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -192,6 +201,74 @@ def __init__(self,
+                for p in param_group['params']:
+                    state = self.state[p]
+                    s = {}
+                    if 'rng_state' not in state:
+                        rng_state = torch.get_rng_state()
+                        s['rng_state'] = rng_state
+                    else:
+                        rng_state = state['rng_state']
+                        torch.set_rng_state(rng_state)
+                    n = p.numel()
+                    ratio = max(self._min_capacity / n, capacity_ratio)
+                    mask = torch.rand_like(p) >= ratio
+                    s['mask'] = mask
```
It seems that mask generation can be put into apply_mask, so that we don't need to save all the masks in state.

Haichao-Zhang replied:
Changed as suggested

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -170,16 +173,22 @@ def __init__(self,
+        capacity_ratio_scheduler = alf.utils.schedulers.as_scheduler(
+            capacity_ratio)
+
+        self._capacity_ratio = capacity_ratio_scheduler()
```
why not use scheduler as the old code?

Haichao-Zhang replied:
Change to the original way of naming

Haichao-Zhang committed:
Address comments

Haichao-Zhang committed:
Fix typo

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -170,16 +173,23 @@ def __init__(self,
+
+        # record initial capacity ratio in order to be used in ``add_param_group``,
+        # within which the capacity scheduler (``self._capacity_ratio``) is inaccessible
+        self._initial_capacity_ratio = capacity_ratio()
```
why not assign self._capacity_ratio?

Haichao-Zhang replied:
Changed. I actually tried this initially when implementing this PR this yesterday, but was unsuccessful. Probably due to other issues at that time.  And that's why I recorded the initial capacity ratio value in the original PR.
Bow this seems unnecessary.

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -192,6 +202,77 @@ def __init__(self,
+                    # get, save and set random number generator state
+                    if 'rng_state' not in state:
+                        rng_state = torch.get_rng_state()
+                        s['rng_state'] = rng_state
```
`state['rng_state'] = rng_state` and line 270-271 can be removed.

Haichao-Zhang replied:
Changed as suggested

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -192,6 +202,77 @@ def __init__(self,
             NewCls.counter += 1
 
+    @common.add_method(NewCls)
+    def _prepare_capacity_states(self, capacity_ratio: float):
+        """prepare capacity related states
+
+        Args:
+            capacity_ratio: the capacity ration specifying the ratio of learnable parameters
+                to the number of all parameters
+        Returns:
+            states: the states containing:
+                - old_param: the value of the parameters before gradient step if
+                    ``self._masked_out_value`` is not None
+        """
+        states = {}
+        if capacity_ratio < 1:
+            for param_group in self.param_groups:
+                for p in param_group['params']:
+                    s = {}
+                    if self._masked_out_value is None:
+                        # only save previous param value if masked_out_value is unspecified
+                        s['old_param'] = p.data.clone()
+                    states[p] = s
+        return states
```
Given the current implementation, it's better to change this to `_clone_params()` and return a dict where the value is the cloned parameter.

Haichao-Zhang replied:
Changed function name and other relevant places accordlingly.

Haichao-Zhang committed:
Address comments

Haichao-Zhang committed:
minor update

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -192,6 +198,79 @@ def __init__(self,
+        if capacity_ratio < 1:
+            for param_group in self.param_groups:
+                for p in param_group['params']:
+                    if self._masked_out_value is None:
```
combine with line 215

Haichao-Zhang replied:
Yes, it should be combined. Done

Haichao-Zhang committed:
Address further comments

Haichao-Zhang committed:
Fix rng_state bug

Haichao-Zhang commented:
Update: now behavior is normal, when using masked_out_values=0: 
<img width="223" alt="image" src="https://github.com/HorizonRobotics/alf/assets/21375027/6d700c3b-017b-4704-947a-4012ed03dd7b">


The abnormal behavior (very bad performance) I mentioned days ago leads to the discovery of the bug fixed in 4c4ac73f80365f3b868e2c83c16d13ba2e0b8ad3. 
The issue was that previously we only save and recover the torch cpu rng_state, therefore when running on GPU, the mask for the same parameter in each optimizer step is **different**.  Therefore, in the current main branch, the capacity scheduling actually leads to an effects of stopping gradient on a subset of randomly selected parameter elements per gradient step. 
In the case of using masked_out_value, the randomly selected parameter elements with be set to zero value. Because the selected subset is random each time even if the capacity ratio is the same, this will leads to a bigger change to the effective function represented by the network, resulting bad and strange performance.

Haichao-Zhang committed:
Address comments and add unittest

Haichao-Zhang commented:
Update II: addressed comments in 14b29809a2973f0ab93d866b44a8433727689bb7.
1. paired the rng_state with device obtained from ``alf.get_default_device``, and also use the corresponding device when setting rng_state. By doing this, we can avoid introducing additional member attributes to optimizer.
2. added unit test. Tested locally on both cpu and gpu the test cases passed.

Haichao-Zhang committed:
Add rng as member attribute

Haichao-Zhang commented:
Now saved a random number generator as member attribute 79f24f6f2fb98558d9eb06fe84215df0d18f0286

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -192,6 +201,80 @@ def __init__(self,
+                    # get, save and set random number generator state
+                    if 'rng_state' not in state:
+                        # record random number generator state in ``self.state``
+                        state['rng_state'] = common.get_torch_rng_state(
+                            self._random_number_generator.device)
```
`self._random_number_generator.get_state()`

Haichao-Zhang replied:
Fixed

emailweixu commented on alf/optimizers/optimizers.py:
```diff
@@ -192,6 +201,80 @@ def __init__(self,
+                    else:
+                        rng_state = state['rng_state']
+                        common.set_torch_rng_state(*rng_state)
+
```
`self._random_number_generator.set_state()`

I wonder how the unittest can pass.

Haichao-Zhang replied:
> `self._random_number_generator.set_state()`
> 
> I wonder how the unittest can pass.

The reason was that the unittests are designed for the previous version before using rnd as member attributes. 
Updated the unittests and also in optimizers.

This issue only present in https://github.com/HorizonRobotics/alf/commit/79f24f6f2fb98558d9eb06fe84215df0d18f0286, previous results (e.g. the return curves were obtained before introducing this commit, and does not have this issue)

Haichao-Zhang committed:
use local_rng

emailweixu commented on alf/optimizers/optimizers_test.py:
```diff
@@ -252,6 +253,64 @@ def _test(i, sampled_predictive=False):
+
+        rand_matrix0 = _generate_rand_matrix_for_capacity_mask(opt)
+
+        opt.zero_grad()
```
Instead of comparing with a mask generated outside of opt. It's better to infer the mask from the parameter before and after the backward() call. And the mask should not change with multiple backward() calls.

Haichao-Zhang replied:
Good suggestion. Done

Haichao-Zhang committed:
update tests with capacity mask inferred from param pairs

Haichao-Zhang committed:
minor update

emailweixu commented on alf/optimizers/optimizers_test.py:
```diff
@@ -21,7 +22,8 @@
 from alf.tensor_specs import TensorSpec
-from alf.utils import tensor_utils
+from alf.utils import common, tensor_utils
+from alf.utils.schedulers import LinearScheduler
```
not used

emailweixu commented on alf/optimizers/optimizers_test.py:
```diff
@@ -252,6 +254,64 @@ def _test(i, sampled_predictive=False):
+        dict(capacity_ratio=0.5, masked_out_value=0, opt_steps=5))
+    def test_capacity_scheduling(self, capacity_ratio, masked_out_value,
+                                 opt_steps):
+        layer = torch.nn.Linear(5, 3)
```
use a large parameter and check the mask is approximately close to the desired capacity.

Haichao-Zhang replied:
Added. A constant capacity is used to keep the test not too complicated.

Haichao-Zhang committed:
Update unittest
